File: android_app/app/src/main/java/com/example/phonecursor/SecureSession.kt
package com.example.phonecursor

/**
 * Owns the ECDH handshake and derived AES-256-GCM state for one
 * phone<->desktop session, from the phone's point of view (mirrors
 * desktop_server's SecureSession.h/.cpp; see the SecureEnvelope docs in
 * Protocol.kt for the wire format this produces/consumes).
 *
 * The phone always initiates with a HELLO and then waits: for a Local
 * connection the desktop confirms automatically, for an Internet
 * connection the operator must compare [confirmationCode] against the
 * desktop's console and accept it there first.
 */
class SecureSession {
    enum class Phase {
        AWAITING_HELLO_ACK,
        AWAITING_CONFIRM,
        CONFIRMED,
        REJECTED,
    }

    var phase: Phase = Phase.AWAITING_HELLO_ACK
        private set

    var confirmationCode: Int = 0
        private set

    private val localKeyPair = Crypto.EcdhKeyPair()
    private val localNonce = Crypto.randomBytes(Crypto.NONCE_SIZE)

    // phone -> desktop
    private lateinit var sendSalt: ByteArray
    private lateinit var sendKey: ByteArray
    private var sendCounter = 0L

    // desktop -> phone
    private lateinit var receiveSalt: ByteArray
    private lateinit var receiveKey: ByteArray
    private var highestReceivedCounter = -1L
    private var hasReceivedAny = false

    fun buildHello(): ByteArray = Protocol.hello(Protocol.TYPE_HELLO, localKeyPair.localPublicKey, localNonce)

    /** Consumes the desktop's HELLO_ACK: derives session key material and the confirmation code. */
    fun handleHelloAck(helloAck: Protocol.HelloPacket) {
        val sharedSecret = localKeyPair.agreeWith(helloAck.publicKey)
        val hkdfSalt = localNonce + helloAck.nonce // phoneNonce || desktopNonce, matches desktop's derivation

        val codeBytes = Crypto.hkdf(sharedSecret, hkdfSalt, Config.HKDF_INFO_CONFIRMATION_CODE, 4)
        val codeValue = ((codeBytes[0].toInt() and 0xFF).toLong() shl 24) or
            ((codeBytes[1].toInt() and 0xFF).toLong() shl 16) or
            ((codeBytes[2].toInt() and 0xFF).toLong() shl 8) or
            (codeBytes[3].toInt() and 0xFF).toLong()
        var modulus = 1L
        repeat(Config.CONFIRMATION_CODE_DIGITS) { modulus *= 10 }
        confirmationCode = (codeValue % modulus).toInt()

        val (sSalt, sKey) = deriveDirectionKey(sharedSecret, hkdfSalt, Config.HKDF_INFO_PHONE_TO_DESKTOP_KEY)
        sendSalt = sSalt
        sendKey = sKey

        val (rSalt, rKey) = deriveDirectionKey(sharedSecret, hkdfSalt, Config.HKDF_INFO_DESKTOP_TO_PHONE_KEY)
        receiveSalt = rSalt
        receiveKey = rKey

        phase = Phase.AWAITING_CONFIRM
    }

    fun markConfirmed() {
        phase = Phase.CONFIRMED
    }

    fun markRejected() {
        phase = Phase.REJECTED
    }

    /** Wraps an outgoing phone->desktop packet ([Protocol]'s builders' output: type byte + payload)
     * into a SecureEnvelope. */
    fun encrypt(fullPacket: ByteArray): ByteArray {
        val type = fullPacket[0]
        val payload = fullPacket.copyOfRange(1, fullPacket.size)
        val counter = sendCounter++
        val nonce = buildNonce(sendSalt, counter)

        return if (Protocol.requiresConfidentiality(type)) {
            val sealed = Crypto.seal(sendKey, nonce, byteArrayOf(type), payload)
            Protocol.buildSecureEnvelope(type, counter, sealed.tag, sealed.ciphertext)
        } else {
            val sealed = Crypto.seal(sendKey, nonce, byteArrayOf(type) + payload, ByteArray(0))
            Protocol.buildSecureEnvelope(type, counter, sealed.tag, payload)
        }
    }

    /** Unwraps an incoming SecureEnvelope from the desktop. Returns the reconstructed packet
     * bytes (type byte + payload), or null if malformed, unauthenticated, or a replay. */
    fun decrypt(data: ByteArray): ByteArray? {
        val envelope = Protocol.parseSecureEnvelope(data) ?: return null
        if (hasReceivedAny && envelope.counter <= highestReceivedCounter) return null // replay

        val nonce = buildNonce(receiveSalt, envelope.counter)
        val plaintext = if (Protocol.requiresConfidentiality(envelope.innerType)) {
            Crypto.open(receiveKey, nonce, byteArrayOf(envelope.innerType), envelope.body, envelope.tag)
        } else {
            val aad = byteArrayOf(envelope.innerType) + envelope.body
            Crypto.open(receiveKey, nonce, aad, ByteArray(0), envelope.tag)?.let { envelope.body }
        } ?: return null

        highestReceivedCounter = envelope.counter
        hasReceivedAny = true
        return byteArrayOf(envelope.innerType) + plaintext
    }

    private fun deriveDirectionKey(sharedSecret: ByteArray, hkdfSalt: ByteArray, info: String): Pair<ByteArray, ByteArray> {
        val material = Crypto.hkdf(sharedSecret, hkdfSalt, info, Crypto.AEAD_SALT_SIZE + Crypto.AEAD_KEY_SIZE)
        val salt = material.copyOfRange(0, Crypto.AEAD_SALT_SIZE)
        val key = material.copyOfRange(Crypto.AEAD_SALT_SIZE, material.size)
        return salt to key
    }

    private fun buildNonce(salt: ByteArray, counter: Long): ByteArray {
        val nonce = ByteArray(Crypto.AEAD_NONCE_SIZE)
        salt.copyInto(nonce, 0)
        for (i in 0 until 8) {
            nonce[salt.size + i] = ((counter shr (8 * (7 - i))) and 0xFF).toByte()
        }
        return nonce
    }
}

File: desktop_server/src/SecureSession.cpp
#include "SecureSession.h"

#include <cstring>

#include "Config.h"

namespace {

uint32_t Pow10(int exponent) {
    uint32_t result = 1;
    for (int i = 0; i < exponent; ++i) result *= 10;
    return result;
}

// HKDF salt binding both sides' handshake nonces together, so the derived
// keys/code are tied to this exact handshake (not just the ECDH secret).
std::vector<uint8_t> HandshakeSalt(const Crypto::Nonce& phoneNonce, const Crypto::Nonce& desktopNonce) {
    std::vector<uint8_t> salt;
    salt.insert(salt.end(), phoneNonce.begin(), phoneNonce.end());
    salt.insert(salt.end(), desktopNonce.begin(), desktopNonce.end());
    return salt;
}

// Derives a direction's AEAD salt + key together from one HKDF call.
void DeriveDirectionKey(const Crypto::SharedSecret& sharedSecret, const std::vector<uint8_t>& hkdfSalt,
                         const std::string& info, Crypto::AeadSalt& outSalt, Crypto::AeadKey& outKey) {
    auto material = Crypto::Hkdf(sharedSecret.data(), sharedSecret.size(), hkdfSalt.data(), hkdfSalt.size(), info,
                                  Crypto::kAeadSaltSize + Crypto::kAeadKeySize);
    std::memcpy(outSalt.data(), material.data(), Crypto::kAeadSaltSize);
    std::memcpy(outKey.data(), material.data() + Crypto::kAeadSaltSize, Crypto::kAeadKeySize);
}

}  // namespace

SecureSession::SecureSession(bool requireConfirmation)
    : requireConfirmation_(requireConfirmation), phase_(Phase::kAwaitingHello) {}

std::vector<uint8_t> SecureSession::HandleHello(const Protocol::HelloPacket& hello) {
    Crypto::RandomBytes(localNonce_.data(), localNonce_.size());

    Crypto::SharedSecret sharedSecret = localKeyPair_.AgreeWith(hello.publicKey);
    std::vector<uint8_t> hkdfSalt = HandshakeSalt(hello.nonce, localNonce_);

    auto codeBytes = Crypto::Hkdf(sharedSecret.data(), sharedSecret.size(), hkdfSalt.data(), hkdfSalt.size(),
                                   Config::kHkdfInfoConfirmationCode, sizeof(uint32_t));
    uint32_t codeValue = (static_cast<uint32_t>(codeBytes[0]) << 24) | (static_cast<uint32_t>(codeBytes[1]) << 16) |
                          (static_cast<uint32_t>(codeBytes[2]) << 8) | codeBytes[3];
    confirmationCode_ = codeValue % Pow10(Config::kConfirmationCodeDigits);

    DeriveDirectionKey(sharedSecret, hkdfSalt, Config::kHkdfInfoDesktopToPhoneKey, sendSalt_, sendKey_);
    DeriveDirectionKey(sharedSecret, hkdfSalt, Config::kHkdfInfoPhoneToDesktopKey, receiveSalt_, receiveKey_);

    phase_ = requireConfirmation_ ? Phase::kAwaitingConfirm : Phase::kConfirmed;
    return Protocol::BuildHello(Protocol::PacketType::kHelloAck, localKeyPair_.LocalPublicKey(), localNonce_);
}

std::vector<uint8_t> SecureSession::Confirm() {
    phase_ = Phase::kConfirmed;
    return Encrypt(Protocol::PacketType::kHandshakeConfirm, nullptr, 0);
}

std::vector<uint8_t> SecureSession::Reject() {
    phase_ = Phase::kRejected;
    return Encrypt(Protocol::PacketType::kHandshakeReject, nullptr, 0);
}

Crypto::AeadNonce SecureSession::BuildNonce(const Crypto::AeadSalt& salt, uint64_t counter) const {
    Crypto::AeadNonce nonce{};
    std::memcpy(nonce.data(), salt.data(), salt.size());
    for (size_t i = 0; i < sizeof(counter); ++i) {
        nonce[salt.size() + i] = static_cast<uint8_t>((counter >> (8 * (sizeof(counter) - 1 - i))) & 0xFF);
    }
    return nonce;
}

std::vector<uint8_t> SecureSession::Encrypt(Protocol::PacketType type, const uint8_t* payload, size_t payloadLen) {
    uint64_t counter = sendCounter_++;
    Crypto::AeadNonce nonce = BuildNonce(sendSalt_, counter);

    std::vector<uint8_t> aad{static_cast<uint8_t>(type)};
    if (Protocol::RequiresConfidentiality(type)) {
        auto sealed = Crypto::Seal(sendKey_, nonce, aad.data(), aad.size(), payload, payloadLen);
        return Protocol::BuildSecureEnvelope(type, counter, sealed.tag, sealed.ciphertext.data(),
                                              sealed.ciphertext.size());
    }

    aad.insert(aad.end(), payload, payload + payloadLen);
    auto sealed = Crypto::Seal(sendKey_, nonce, aad.data(), aad.size(), nullptr, 0);
    return Protocol::BuildSecureEnvelope(type, counter, sealed.tag, payload, payloadLen);
}

std::optional<std::vector<uint8_t>> SecureSession::Decrypt(const uint8_t* data, size_t length) {
    auto envelope = Protocol::ParseSecureEnvelope(data, length);
    if (!envelope) return std::nullopt;
    if (hasReceivedAny_ && envelope->counter <= highestReceivedCounter_) return std::nullopt;  // replay

    Crypto::AeadNonce nonce = BuildNonce(receiveSalt_, envelope->counter);
    std::vector<uint8_t> aad{static_cast<uint8_t>(envelope->innerType)};

    std::optional<std::vector<uint8_t>> plaintext;
    if (Protocol::RequiresConfidentiality(envelope->innerType)) {
        plaintext = Crypto::Open(receiveKey_, nonce, aad.data(), aad.size(), envelope->body, envelope->bodyLength,
                                  envelope->tag);
    } else {
        aad.insert(aad.end(), envelope->body, envelope->body + envelope->bodyLength);
        if (Crypto::Open(receiveKey_, nonce, aad.data(), aad.size(), nullptr, 0, envelope->tag)) {
            plaintext = std::vector<uint8_t>(envelope->body, envelope->body + envelope->bodyLength);
        }
    }
    if (!plaintext) return std::nullopt;

    highestReceivedCounter_ = envelope->counter;
    hasReceivedAny_ = true;

    std::vector<uint8_t> reconstructed;
    reconstructed.reserve(1 + plaintext->size());
    reconstructed.push_back(static_cast<uint8_t>(envelope->innerType));
    reconstructed.insert(reconstructed.end(), plaintext->begin(), plaintext->end());
    return reconstructed;
}
